pkg/dockerfile/fast_generator.go

```diff
@@ -0,0 +1,261 @@
+	if err != nil {
+		return "", err
+	}
+
+	lines, err = g.copyWeights(lines)
+	if err != nil {
+		return "", err
+	}
+
+	lines, err = g.install(lines)
+	if err != nil {
+		return "", err
+	}
+
+	lines, err = g.entrypoint(lines)
+	if err != nil {
+		return "", err
+	}
+
+	return strings.Join(lines, "\n"), nil
+}
+
+func (g *FastGenerator) copyCog(tmpDir string) (string, error) {
+	files, err := CogEmbed.ReadDir("embed")
+	if err != nil {
+		return "", err
+	}
+	if len(files) != 1 {
+		return "", fmt.Errorf("should only have one cog wheel embedded")
+	}
+	filename := files[0].Name()
+	data, err := CogEmbed.ReadFile("embed/" + filename)
+	if err != nil {
+		return "", err
+	}
+	path := filepath.Join(tmpDir, filename)
+	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
+		return "", fmt.Errorf("Failed to write %s: %w", filename, err)
+	}
+	if err := os.WriteFile(path, data, 0o644); err != nil {
+		return "", fmt.Errorf("Failed to write %s: %w", filename, err)
+	}
+	return path, nil
+}
+
+func (g *FastGenerator) generateMonobase(lines []string, tmpDir string) ([]string, error) {
+	lines = append(lines, []string{
+		"FROM r8.im/monobase:latest",
+	}...)
+
+	cogPath, err := g.copyCog(tmpDir)
+	if err != nil {
+		return nil, err
+	}
+
+	lines = append(lines, []string{
+		"ENV R8_COG_VERSION=\"file:///buildtmp/" + filepath.Base(cogPath) + "\"",
+	}...)
+
+	relativeTmpDir, err := filepath.Rel(g.Dir, tmpDir)
+	if err != nil {
+		return nil, err
+	}
+	skipCudaArg := "--skip-cuda"
+	if g.Config.Build.GPU {
+		skipCudaArg = ""
+		cudaVersion := g.Config.Build.CUDA
+		cudnnVersion := g.Config.Build.CuDNN
+		lines = append(lines, []string{
+			"ENV R8_CUDA_VERSION=" + cudaVersion,
+			"ENV R8_CUDNN_VERSION=" + cudnnVersion,
+			"ENV R8_CUDA_PREFIX=https://monobase.replicate.delivery/cuda",
+			"ENV R8_CUDNN_PREFIX=https://monobase.replicate.delivery/cudnn",
+		}...)
+	}
+
+	lines = append(lines, []string{
+		"ENV R8_PYTHON_VERSION=" + g.Config.Build.PythonVersion,
+	}...)
+
+	torchVersion, ok := g.Config.TorchVersion()
+	if ok {
+		lines = append(lines, []string{
+			"ENV R8_TORCH_VERSION=" + torchVersion,
+		}...)
+	}
+
+	return append(lines, []string{
+		"RUN --mount=type=bind,source=\"" + relativeTmpDir + "\",target=/buildtmp /opt/r8/monobase/build.sh " + skipCudaArg + " --mini",
+	}...), nil
+}
+
+func (g *FastGenerator) copyWeights(lines []string) ([]string, error) {
+	weights, err := FindWeights(g.Dir)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(weights) == 0 {
+		return lines, nil
+	}
+
+	commands := []string{}
+	for sha256, file := range weights {
+		rel_path, err := filepath.Rel(g.Dir, file)
+		if err != nil {
+			return nil, err
+		}
+		commands = append(commands, "cp /src/"+rel_path+" "+filepath.Join(FUSE_RPC_WEIGHTS_PATH, sha256))
+	}
+
+	return append(lines, []string{
+		"RUN --mount=type=bind,ro,source=.,target=/src mkdir -p " + FUSE_RPC_WEIGHTS_PATH + " && " + strings.Join(commands, " && "),
```

Comment:
Might be easier and more reliable to weird file names if you do `rsync -a --include-from=weights-list.txt`. We can also bake the list into container FS & as a label for easy retrieval.

Reply:
I guess we need to bake a file with SHA <-> path mapping, so that we can recreate symlinks later in prod. OTOH `rsync`'s include file is just the paths.

Reply:
Ended up going with `COPY --link` on separate layers, makes builds superfast since it doesn't copy the weights, added some more logic to create a manifest then check the files mod time and size against that manifest (to prevent sha256 digests on the big weights files): https://github.com/replicate/cog/pull/2108/commits/23a109a85767d6577246094d301a139e23644d89

Reply:
Ended up going with `COPY --link` on separate layers, makes builds superfast since it doesn't copy the weights, added some more logic to create a manifest then check the files mod time and size against that manifest (to prevent sha256 digests on the big weights files): https://github.com/replicate/cog/pull/2108/commits/23a109a85767d6577246094d301a139e23644d89

Reply:
Didn't know `--link` nice!